### PR TITLE
tenant-manager: audit logging

### PR DESCRIPTION
**Pull request by famartinrh**
this PR adds audit logging to the tenant-manager. I implemented it using an interceptor

It's missing adding auditing for authentication failures

and also missing gather client IPs

### CODE REVIEW

File: multitenancy/tenant-manager-api/src/main/java/io/apicurio/multitenant/logging/audit/AuditedInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+@Audited
+@Interceptor
+@Priority(Interceptor.Priority.APPLICATION)
+public class AuditedInterceptor {
+
+    @Inject
+    AuditLogService auditLogService;
+
+    @AroundInvoke
+    public Object authorizeMethod(InvocationContext context) throws Exception {
```
Comment (EricWittmann) [inline multitenancy/tenant-manager-api/src/main/java/io/apicurio/multitenant/logging/audit/AuditedInterceptor.java:52]: `authorizeMethod`?

Comment (famartinrh) [reply multitenancy/tenant-manager-api/src/main/java/io/apicurio/multitenant/logging/audit/AuditedInterceptor.java:52]: ouch

**Comment by famartinrh** [comment]
regarding client IPs and logging authentication failures a much simpler solution could be to enable access logs in quarkus, which will generate log entries like that
```
2021-10-01 18:49:10,203 INFO  [io.qua.htt.access-log] (vert.x-eventloop-thread-14) 0:0:0:0:0:0:0:1 - - 01/Oct/2021:18:49:10 +0200 "GET /api/tenants HTTP/1.1" 401 -
```
I think access log could be configured to make the log message look like something more similar to the format of the audit logs I added by hand

And regarding client IPs this would be something to investigate https://quarkus.io/guides/http-reference#reverse-proxy

My vote is for making the quarkus access logs look nicer and to have a combination of quarkus access logs and hand made audit logs to cover our auditing needs, and to investigate later if and how we need that reverse-proxy quarkus config, thoughts?

**Comment by EricWittmann** [comment]
The access logs seem useful, although perhaps too verbose?  I think we only want to audit successful changes, not every access.  Although enabling access logs when debugging a problem would be super valuable.

That reverse-proxy config in Quarkus sounds VERY handy.

Let's first see about getting the client IP address added to the by-hand audit logging, and then see if we need the reverse-proxy config based on the result.

**Comment by carlesarnal** [comment]
Looks like a good start to me.

> The access logs seem useful, although perhaps too verbose? I think we only want to audit successful changes, not every access. Although enabling access logs when debugging a problem would be super valuable.

+1. By default, I think we only want to log successful changes but I think we also want to log unsuccessful accesses. I can think of some scenarios where logging unsuccessful accesses could be very helpful so +1 to add them as e.g. debug.

> 
> That reverse-proxy config in Quarkus sounds VERY handy.
> 
> Let's first see about getting the client IP address added to the by-hand audit logging, and then see if we need the reverse-proxy config based on the result.

**Comment by famartinrh** [comment]
I'm having an issue where I'm unable to log unsuccessfull requests due to authentication error, because looks like Quarkus won't execute my filter if authentication fails

**Comment by carlesarnal** [comment]
> I'm having an issue where I'm unable to log unsuccessfull requests due to authentication error, because looks like Quarkus won't execute my filter if authentication fails

Oh, yes, quarkus is using what they call proactive authentication. Your filter is not going to be executed if the request fails to authenticate. Let me investigate since I think that can be disabled or modified.

**Comment by carlesarnal** [comment]
> > I'm having an issue where I'm unable to log unsuccessfull requests due to authentication error, because looks like Quarkus won't execute my filter if authentication fails
> 
> Oh, yes, quarkus is using what they call proactive authentication. Your filter is not going to be executed if the request fails to authenticate. Let me investigate since I think that can be disabled or modified.

Ok, this was an easy investigation, just add `quarkus.http.auth.proactive=false` and your filter will be executed.

**Comment by famartinrh** [comment]
this should be 100% now, I had to create a Custom HttpAuthenticationMechanism in order to be able to handle auth errors

File: multitenancy/tenant-manager-api/src/main/java/io/apicurio/multitenant/auth/CustomAuthenticationMechanism.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * Custom HttpAuthenticationMechanism that simply wraps OidcAuthenticationMechanism.
+ * The only purpose of this HttpAuthenticationMechanism is to handle authentication errors in order to generate audit logs.
+ *
+ * @author Fabian Martinez
+ */
+@Alternative
```
Comment (carlesarnal) [inline multitenancy/tenant-manager-api/src/main/java/io/apicurio/multitenant/auth/CustomAuthenticationMechanism.java:54]: Yes, I forgot this is required, I added it in Registry to the BasicAuthClientCredentialsMechanism and that one is relying on the CustomAuthenticationMechanism, sorry about that.

**Review by carlesarnal — APPROVED** [review]
LGTM. This makes me wonder if we should implement audit logging as a library that can be re-used across components since I'm going to implement something very similar for Registry and we will probably end up implementing something similar for fleet-manager. Something to discuss in the future, probably.

**Comment by famartinrh** [comment]
just to avoid even more productization work I would say, not to create a common library, and if so only have it for apicurio-registry and tenant-manager, so we avoid releasing the library.

you ok to merge this? and later we can extract some common logic to another module

**Pull request merged**